design/open/lakefs-kv-execution-plan.md

```diff
@@ -0,0 +1,72 @@
+# lakeFS on KV - Execution Plan
+
+Key/value data storage package (kv) that will supply key/value access to data to replace the current storage and locking done by Postgres.
+High level overview of the interface described in https://github.com/treeverse/lakeFS/blob/ed55edc2c24b24ed4e6fcccad1ae95844d38b9ad/design/open/metadata_kv/index.md. The package will support multiple back-end implementations.
+
+## Transform from using db to kv package
+
+High level interface described in the kv proposal, we should consider taking the DynamoDB interface to address the way we communicate `set-if` functionality while marshal the same value without fetching data first.
+
+Key/value format can be done using ProtoBuf / JSON.
+
+- key format based on path - a format with escaping should be supported to enable easy way to encode values as part of the key
+- each model should include version as part of the model information
+- migrate from old model to new one without downtime
+
+## Per package changes
+
+For the following packages we should handle to move from 'db' package to 'kv'.
+During development it should enable a configuration flag to switch between using kv and db implementation.
+Any move transition should support migration from old Postgres data to new Postgres data format.
+When a complete data model / functionality completes the transition we can exclude the 'option' to swtich from/to and just use the new format.
+
+- Map the use of transactions in order to provide a valid solution using the kv interface
+- Map any database locking to specific solution using the kv implementation
```

Comment:
I think we need to expand on these two and offer alternatives. It basically offers no help in the kv-world, and I'm sure we can come up with more primitives on top of the existing interface that can set a common ground to solve most of the common issues around locking, transactions, secondary indexes, etc.. If each package implementor is left to build her own on top of the kv interface, we'll end up with a much more complex solution.

Reply:
The database locking alternative is based on set-if implementation in the current proposal - updated the text.
About the transaction - we listed couple of options: using set-if, check secondary index after reading etc.
@itaiad200 do you think we need to list the ones we discussed on the first item as possible options, or it should be part of the task / design?